### PR TITLE
Add producer buffer flushing option

### PR DESCRIPTION
**Pull request by arabelle**
Once the Kafka Connect producer buffer is full, the consumer thread blocks until there is enough space to allocate a new buffer for whatever partition it happens to have a record for. Eventually it allocates a new buffer, but the next record will be allocated to a different partition, and thus is likely to block the consumer again waiting for another buffer to be allocated. This may keep happening, and by the time there is more data available for the first partition that particular buffer is likely to have reached the linger time - so the Sender will send the batch, even though it may contain only one or two records. At this point the problem snowballs as we start sending lots of small inefficiently compressed batches making blocking even more likely. In Kafka Connect we keep trying to consume continuously, so the buffer is never fully cleared. Add a config option to enable consumer blocking until the buffer empties.

### CODE REVIEW

File: src/main/java/com/salesforce/mirus/MirusSourceTask.java [hunk 1]
```diff
@@ -151,6 +156,12 @@ private void seekToOffsets(List<TopicPartition> partitionIds) {
   public List<SourceRecord> poll() {
 
     try {
+      if (this.enableBufferFlushing) {
```
Comment (ddbrodsky) [inline src/main/java/com/salesforce/mirus/MirusSourceTask.java]: Combine these into a single conditional statement?

File: src/main/java/com/salesforce/mirus/MirusSourceTask.java [hunk 2]
```diff
@@ -162,6 +173,8 @@ private void seekToOffsets(List<TopicPartition> partitionIds) {
     } catch (WakeupException e) {
       // Ignore exception iff shutting down thread.
       if (!shutDown.get()) throw e;
+    } catch (Exception re) {
```
[the remark this answers is not in the document]
Comment (arabelle) [reply src/main/java/com/salesforce/mirus/MirusSourceTask.java]: Good point, I overlooked that when I was cleaning some code up.

File: src/main/java/com/salesforce/mirus/MirusSourceTask.java [hunk 3]
```diff
@@ -160,8 +171,10 @@ private void seekToOffsets(List<TopicPartition> partitionIds) {
         return Collections.emptyList();
       }
     } catch (WakeupException e) {
-      // Ignore exception iff shutting down thread.
```
Comment (ddbrodsky) [inline src/main/java/com/salesforce/mirus/MirusSourceTask.java]: nitpick --  iff may have been short for if and only if  :)

Comment (arabelle) [reply src/main/java/com/salesforce/mirus/MirusSourceTask.java]: True true :)

File: src/main/java/com/salesforce/mirus/metrics/QueryJmxBean.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+  public boolean bufferAvailable() throws InstanceNotFoundException, ReflectionException {
+    Set<ObjectName> producerMetrics = new HashSet<>();
+    try {
+      logger.trace("Querying producer metrics for each client-id.");
+      producerMetrics =
+          server.queryNames(
+              new ObjectName("kafka.producer:type=producer-metrics,client-id=*"), null);
+    } catch (MalformedObjectNameException e) {
+      logger.error(e.getMessage());
```
Comment (ddbrodsky) [inline src/main/java/com/salesforce/mirus/metrics/QueryJmxBean.java]: At this point here, can we proceed, or do we have to abort this check?

File: src/main/java/com/salesforce/mirus/metrics/QueryJmxBean.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    }
+    logger.trace("Got {} producers", producerMetrics.size());
+    for (ObjectName producer : producerMetrics) {
+      Double bufferAvailableBytes = 0.0;
+      try {
+        // Buffer available bytes is of type Double
+        bufferAvailableBytes = (Double) server.getAttribute(producer, "buffer-available-bytes");
+      } catch (MBeanException | AttributeNotFoundException e) {
+        logger.error(e.getMessage());
```
Comment (ddbrodsky) [inline src/main/java/com/salesforce/mirus/metrics/QueryJmxBean.java]: What happens to the value of `bufferAvailableBytes` if we error out?

**Pull request closed without merging**